### DISCUSSION TITLE
Episode 196: Runtime warning work only in half of the cases

### DISCUSSION THREAD

**Discussion by vladvlasov256** [pointfreeco/episode-code-samples#123] [General]
**Episode**: [196](https://www.pointfree.co/episodes/ep196-async-composable-architecture-tasks)
**Time**: 35:20

I've noticed that runtime warning in the `equals()` in the [File.swift](https://github.com/pointfreeco/episode-code-samples/blob/7f5b443dc4d84b22d5972c6b4a4c01ebae266e8d/0196-tca-concurrency-pt2/swift-composable-architecture/Sources/ComposableArchitecture/File.swift) is fired only if `lhs` is non-equatable.

But what if `rhs` is non-equable too? Or `rhs` is the only one which is non-equatable here.

```
public func equals(_ lhs: Any, _ rhs: Any) -> Bool {
  ...

  guard
    let lhs = lhs as? any Equatable
  else {
    runtimeWarning(
      "Tried to compare a non-equatable error type: %@",
      ["\(type(of: lhs))"]
    )
    return false
  }

  ...
}
```

IMO, a version on 34th minute with two casts is more helpful for users. Or am I wrong?

**Comment by mbrandonw** [comment]
Hi @vladvlasov256, sorry we missed this question. 

The reason it's only needed to check `lhs` is because there's only one place that `equals` is called, and in that case `lhs` are of the same type. Even though technically `lhs` and `rhs` could be different types by virtue of the fact that they are `Any`s, in practice that's not really the case. So it is sufficient to only warn if `lhs` is not equatable.
